### ISSUE TITLE
Add info about performance in the README

### ISSUE THREAD

**Issue by artem-zinnatullin** [pushtorefresh/storio#427]
We can also make some benchmarks like this https://github.com/daj/android-orm-benchmark (not sure about their quality tho)

**Comment by Rainer-Lang** [comment]
Very interesting lib! 
+1 for infos about performance

**Comment by artem-zinnatullin** [comment]
Related to #489.

**Comment by skrzyneckik** [comment]
@artem-zinnatullin in case you didn't know, @touchlab updated released updated [android-orm-benchmark-updated repository](https://github.com/touchlab/android-orm-benchmark-updated) some time ago. I thought about adding StorIO to it as well, but this benchmark is purely for ORM libraries, so imho shouldn't be there. But it didn't prevent me from making a fork of it: [android-db-libraries-benchmark](https://github.com/skrzyneckik/android-db-libraries-benchmark/tree/add_storio)

My concert is that reading data seems to be fast, but writing is really slow. Any ideas?

**Comment by artem-zinnatullin** [comment]
You're seeing slow writes with StorIO or it's just hypothetical? Because we
give you full control over SQL and transactions so the only slow down from
us is few allocations for abstractions like Query and Transaction
notifications.

On 29 Sep 2016 5:49 pm, "Krzysztof Skrzynecki" <notifications@github.com>
wrote:

@artem-zinnatullin <https://github.com/artem-zinnatullin> in case you
didn't know, @touchlab <https://github.com/touchlab> updated released
updated android-orm-benchmark-updated repository
<https://github.com/touchlab/android-orm-benchmark-updated> some time ago.
I thought about adding StorIO to it as well, but this benchmark is purely
for ORM libraries, so imho shouldn't be there. But it didn't prevent me
from making a fork of it: android-db-libraries-benchmark
<https://github.com/skrzyneckik/android-db-libraries-benchmark/tree/add_storio>

My concert is that reading data seems to be fast, but writing is really
slow. Any ideas?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/pushtorefresh/storio/issues/427#issuecomment-250601289>,
or mute the thread

**Comment by skrzyneckik** [comment]
@artem-zinnatullin It's slow in benchmark: put 2000 Users (3 field object) and 20000 Messages (9 fields object) and I do it like that:
```
try {
    storIOSQLite.lowLevel().beginTransaction();

    storIOSQLite.put()
        .objects(users)
        .prepare()
        .executeAsBlocking();

    storIOSQLite.put()
        .objects(messages)
        .prepare()
        .executeAsBlocking();

    storIOSQLite.lowLevel().setTransactionSuccessful();
} catch (Exception e){
    e.printStackTrace();
} finally {
    storIOSQLite.lowLevel().endTransaction();
}
```

Is there anything in particular what I do wrong?

**Comment by artem-zinnatullin** [comment]
Can you please give link to code, I wasn't been able to find StorIO in your
benchmark yesterday (I'm from phone, so that might be a reason), thanks!

**Comment by skrzyneckik** [comment]
@artem-zinnatullin [Here](https://github.com/skrzyneckik/android-db-libraries-benchmark/tree/add_storio/ORM-Benchmark/src/main/java/com/littleinc/orm_benchmark/storio) you can find whole StorIO part package: User and Message represent entities, DataBaseHelper extends SQLiteOpenHelper and StorIOExecutor executes different actions (create table, insert, read and delete table)

**Comment by artem-zinnatullin** [comment]
@skrzyneckik ah, it's on separate branch!

So, I see few problems in your benchmark:

`writeWholeData()`:

1. You use `LinkedList` which gives significant degradation in reading in compare to `ArrayList` since you have to traverse through heap each time you need to get reference to new element, especially on scale you use it at (`NUM_USER_INSERTS = 2000; NUM_MESSAGE_INSERTS = 20000;`).
2. You can pass all items (of different types) as single list to `storIO.put().objects()`, this will decrease number of operations StorIO does under the hood.
3. You use transactions, but you do it in non-great way. By default `storIO.put().objects()` already uses transactions so in your example you will have nested transactions which are not required. In fact you can remove manual transaction if you'll pass all items (of different types) to single `storIO.put().objects()` call, it'll handle transaction for you.

`readWholeData()`:

* I would exclude logging from measuring time.

Otherwise looks interesting! Pls share results after applying changes I mentioned :)

**Comment by skrzyneckik** [comment]
@artem-zinnatullin 
I agree that logging shouldn't be part of measure action, but rest of executors do it in this way and I wanted to just add StorIO to existing benchmarks and be consistent across all of them without modifying old ones

**Comment by skrzyneckik** [comment]
@artem-zinnatullin due to `writeWholeData` I've checked how does it work with changes which you recommended and it's 1.5% faster now, but it's still slow. I'll try to analyse what's going on under the hood and maybe it'll help me to understand what potential bottleneck is.

**Comment by artem-zinnatullin** [comment]
@skrzyneckik how many times do you run the benchmark, is SQLite used as in-memory db or it stores info on disk, on which device/emulator do you run it?

**Comment by skrzyneckik** [comment]
@artem-zinnatullin I found main cause of low `writeWholeData` performance - put operation handles "insert/update if item already exist" and it's a main advantage of StorIO: Put Get Delete instead of CRUD. I completely forgot about it. 

Just for the sake of benchmark, I've replaced generated `DefaultPutResolver` with custom one, which `performPut` method doesn't check that item already exists. Now results look fine: insert all of items with raw sql queries took 2428ms and storIO made it in 2836ms
![screen shot 2016-10-09 at 12 47 25](https://cloud.githubusercontent.com/assets/1184480/19220167/dcfb0a60-8e1e-11e6-95b6-81adae0e40ee.jpg)

I didn't stop there. As you can see raw optimized sql can make it two time faster, because it uses precompiled SQLStatement. To apply this optimalization to StorIO I needed to apply new `LowLevel` implementation to `DefaultStorIOSQLite`. It allows to decrease `writeWholeData` operation to 1850ms.
![screen shot 2016-10-09 at 13 01 33](https://cloud.githubusercontent.com/assets/1184480/19220233/8afa4dd2-8e20-11e6-95ee-d018f5387b7d.jpg)

I can imagine that further optimizations could be applied - e.g. the way I've applied prepared statement to StorIO can't be called 'nice' ;) Moreover I think it doesn't make sense to use brute force to fit StorIO into existing benchmark, which tests how fast are simple insert and get operations. In my opinion to check StorIO performance, benchmark should check 'insert or update' operation instead. Moreover would it possible to fit `SQLStatement` into StorIO somehow?

**Comment by artem-zinnatullin** [comment]
@skrzyneckik 

>put operation handles "insert/update if item already exist" and it's a main advantage of StorIO

This is true, but it's only behavior of `DefaultPutResolver`, user can always write own resolver for Put Operation. By default we do very safe but not very performant Put Operation: we first try to `query` entry with same id/etc and only then do `insert` or `update`. We could use `insertOrUpdate` which is faster, but it creates several constraints for `id`s which we don't want all users to stick to.

>In my opinion to check StorIO performance, benchmark should check 'insert or update' operation instead.

I definitely like this and other conclusions you made, such thinking is pretty rare in becnhmarking world 👍 

>would it possible to fit SQLStatement into StorIO somehow?

Yes, we can add this to `LowLevel` API!

I think your work here is basically the asnwer to the "Performance" question of StorIO — by default it's not fastest DB wrapper, but we give users all high and low-level APIs to improve hot parts of their DB interactions, we're also open for PRs that improve performance without breaking StorIO for other users (though I'm not sure we've left anything except `insertOrReplace()`).

@skrzyneckik thank you for detailed investigation and feedback!

**Comment by skrzyneckik** [comment]
@artem-zinnatullin just for a record: `SQLiteStatement` is ["not thread-safe"](https://developer.android.com/reference/android/database/sqlite/SQLiteStatement.html). So LowLevel should handle it somehow. Later I'll add here some concern which I have about api, which I faced when I tried to replace LowLevel implementation.

**Comment by artem-zinnatullin** [comment]
@skrzyneckik ok, thanks! Until we have requests to add `SQLiteStatement` from users we won't do it to not bloat API with unused functions

**Comment by Rainer-Lang** [comment]
@skrzyneckik What do you think ```SQLiteStatement``` will affect the performance? Do you have any experiences?

**Comment by nikitin-da** [comment]
@artem-zinnatullin there is request for such api from @karlicoss

**Comment by artem-zinnatullin** [comment]
@skrzyneckik if you have benchmark results with and without `SQLiteStatement` (related or not related to StorIO), I'd be happy to see them (with code of course!).

Last time @StanisVS and I played with it I got impression that it doesn't actually add much perf, I guess, SQLite may cache queries on its own.

**Comment by skrzyneckik** [comment]
@artem-zinnatullin to compare benchmark of raw sql query with and without `SQLiteStatement` you can take a look on an [origin orm benchmark](https://github.com/touchlab/android-orm-benchmark-updated), which I used to compare StorIO. Unoptimized raw sql queries can be found [here](https://github.com/touchlab/android-orm-benchmark-updated/tree/master/ORM-Benchmark/src/main/java/com/littleinc/orm_benchmark/sqlite) and optimized [here](https://github.com/touchlab/android-orm-benchmark-updated/tree/master/ORM-Benchmark/src/main/java/com/littleinc/orm_benchmark/sqliteoptimized). It's also available in [Google Play Store](https://play.google.com/store/apps/details?id=com.littleinc.orm_benchmark). Just run it and compare `RAW` results with `RAW_OPTIMIZED`. 

@Rainer-Lang However as I mentioned earlier this benchmark covers only simple scenario when 22000 inserts (no updates) are applied in single transaction and that's the reason why optimized raw sql benchmark is two time faster than unoptimized one - statement is precompiled and it's reused across all of inserts. I would say most of users won't need it because we won't get to much by using `SQLiteStatement` when there are just few thousands items inserted into db.

Moreover, `SQLiteStatement` has few extra requirements - e.g. it can't be compiled until table is created etc.

In my opinion, as I've noticed trying to apply `SQLiteStatement`, `DefaultStorIOSQLite` could be a little more flexible with replacing LowLevel implementation - I'll create a issue for it at some point just to discuss does it make sense to change this API in the first place.

Btw I found whole Master thesis about [An Experimental Study of the Performance, Energy, and Programming Effort Trade-offs of Android Persistence Frameworks](https://vtechworks.lib.vt.edu/bitstream/handle/10919/72268/Pu_J_T_2016.pdf?sequence=1), which can be useful to decide to we need to use this `SQLiteStatement` in the first place.

**Comment by skrzyneckik** [comment]
@Rainer-Lang [here](http://stackoverflow.com/a/1703234/2004186) is what I found:
> Prepared statements improve performance by caching the execution plan for a query after the query optimizer has found the best plan.
So actually it make sense to have it for PutResolver - it could be lazily initialized, but of course we'll need to have a proper benchmark for it. Another option which I see would be to have `SQLiteStatement` next to raw sql - we could decide does it make sense to compile it based on number of put items. As far as I know default implementation of sorting algorithms apply this approach, because of there different performance due to size of data which need to be sorted.

**Comment by skrzyneckik** [comment]
@Rainer-Lang @artem-zinnatullin just to let you know I reviewed `SQLite.insertWithOnConflict` method in Android source ([here](https://gist.github.com/skrzyneckik/54e6a829f57f1aeb1d0036388a1ee51f) you can find gist with it) and you can see there that by default `SQLiteStatement` is compiled for every insert operation anyway, so we'll definitely get some improvement in performance if we'll it do it only once in advance. Moreover, for for the record - `acquireReference` and `releaseReference` are used to synchronize `SQLiteDatabase` object and their implementation doesn't seem too complicated

**Comment by artem-zinnatullin** [comment]
@skrzyneckik great info, thanks! 

@nikitin-da let's investigate how we can use `SQLiteStatement` in our Resolvers under the hood!
